src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncDisposableTests.cs

```diff
@@ -0,0 +1,1192 @@
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "body");
+        }
+
+        [Fact]
+        public void TestAwaitExpressionInfo_IEquatable()
+        {
+            string source = @"
+public class C
+{
+    void GetAwaiter() { }
+    void GetResult() { }
+    bool IsCompleted => true;
+}
+public class D
+{
+    void GetAwaiter() { }
+    void GetResult() { }
+    bool IsCompleted => true;
+}
+";
+            var comp = CreateStandardCompilation(source);
+            var getAwaiter1 = (MethodSymbol)comp.GetMember("C.GetAwaiter");
+            var isCompleted1 = (PropertySymbol)comp.GetMember("C.IsCompleted");
+            var getResult1 = (MethodSymbol)comp.GetMember("C.GetResult");
+            var first = new AwaitExpressionInfo(new AwaitableInfo(getAwaiter1, isCompleted1, getResult1));
+
+            var nulls1 = new AwaitExpressionInfo(new AwaitableInfo(null, isCompleted1, getResult1));
+            var nulls2 = new AwaitExpressionInfo(new AwaitableInfo(getAwaiter1, null, getResult1));
+            var nulls3 = new AwaitExpressionInfo(new AwaitableInfo(getAwaiter1, isCompleted1, null));
+
+            Assert.False(first.Equals(nulls1));
+            Assert.False(first.Equals(nulls2));
+            Assert.False(first.Equals(nulls3));
+
+            Assert.False(nulls1.Equals(first));
+            Assert.False(nulls2.Equals(first));
+            Assert.False(nulls3.Equals(first));
+
+            object nullObj = null;
+            Assert.False(first.Equals(nullObj));
+
+            var getAwaiter2 = (MethodSymbol)comp.GetMember("D.GetAwaiter");
+            var isCompleted2 = (PropertySymbol)comp.GetMember("D.IsCompleted");
+            var getResult2 = (MethodSymbol)comp.GetMember("D.GetResult");
+            var second1 = new AwaitExpressionInfo(new AwaitableInfo(getAwaiter2, isCompleted1, getResult1));
+            var second2 = new AwaitExpressionInfo(new AwaitableInfo(getAwaiter1, isCompleted2, getResult1));
+            var second3 = new AwaitExpressionInfo(new AwaitableInfo(getAwaiter1, isCompleted1, getResult2));
+            var second4 = new AwaitExpressionInfo(new AwaitableInfo(getAwaiter2, isCompleted2, getResult2));
+
+            Assert.False(first.Equals(second1));
+            Assert.False(first.Equals(second2));
+            Assert.False(first.Equals(second3));
+            Assert.False(first.Equals(second4));
+
+            Assert.False(second1.Equals(first));
+            Assert.False(second2.Equals(first));
+            Assert.False(second3.Equals(first));
+            Assert.False(second4.Equals(first));
+
+            Assert.True(first.Equals(first));
+            Assert.True(first.Equals((object)first));
+
+            Assert.True(first.GetHashCode() != 0);
+            Assert.True(first.GetHashCode() != second1.GetHashCode());
```

Comment:
>first.GetHashCode() != second1.GetHashCode() [](start = 24, length = 44)

Usually, it is not required to produce not equal hash codes for not equal instances. The only requirement is to produce equal hash for equal instances.

Reply:
I understand. However I think it's reasonable to check that. If those checked failed, it would be a very likely indication of broken logic inside `GetHashCode`. Ok to leave?

Reply:
> If those checked failed, it would be a very likely indication of broken logic inside GetHashCode. 

Not necessary. This could be a source of flakiness for the test as the value could change from run to run. Instead, we should be testing the requirement that does exists, i.e. equal values produce the same hash. It doesn't look like we test this here. 


---
In reply to: [161612658](https://github.com/dotnet/roslyn/pull/23961#discussion_r161612658) [](ancestors = 161612658)